api/task_manager.cc

```diff
@@ -0,0 +1,178 @@
+    res.parent_id = status.parent_id.to_sstring();
+    res.sequence_number = status.task_status.sequence_number;
+    res.shard = status.task_status.shard;
+    res.keyspace = status.task_status.keyspace;
+    res.table = status.task_status.table;
+    res.entity = status.task_status.entity;
+    res.progress_units = status.task_status.progress_units;
+    res.progress_total = status.progress.total;
+    res.progress_completed = status.progress.completed;
+    return res;
+}
+
+future<json::json_return_type> retrieve_status(task_manager::foreign_task_ptr task) {
+    if (task.get() == nullptr) {
+        co_return coroutine::return_exception(httpd::bad_param_exception("Task not found"));
+    }
+    auto progress = co_await task->get_progress();
+    full_task_status s;
+    s.task_status = task->get_status();
+    s.parent_id = task->get_parent_id();
+    s.abortable = task->is_abortable();
+    s.module = task->get_module()->get_name();
+    s.progress.completed = progress.completed;
+    s.progress.total = progress.total;
+    co_return make_status(s);
+}
+
+void set_task_manager(http_context& ctx, routes& r) {
+    tm::get_modules.set(r, [&ctx] (std::unique_ptr<request> req) -> future<json::json_return_type> {
+        std::vector<std::string> v = boost::copy_range<std::vector<std::string>>(ctx.tm.local().get_modules() | boost::adaptors::map_keys);
+        co_return v;
+    });
+
+    tm::get_tasks.set(r, [&ctx] (std::unique_ptr<request> req) -> future<json::json_return_type> {
+        using chunked_stats = std::vector<task_stats>;
+        std::vector<chunked_stats> res = co_await ctx.tm.map([&req] (task_manager& tm) {
+            chunked_stats local_res;
+            auto module = tm.find_module(req->param["module"]);
+            const auto& filtered_tasks = module->get_tasks() | boost::adaptors::filtered([&params = req->query_parameters] (const auto& task) {
+                return filter_tasks(task.second, params);
+            });
+            for (auto& [task_id, task] : filtered_tasks) {
+                task_stats ts;
+                ts.task_id = task->id().to_sstring();
+                ts.state = task->get_status().state;
+                local_res.push_back(ts);
+            }
+            return local_res;
+        });
+
+        std::function<future<>(output_stream<char>&&)> f = [_res = std::move(res)] (output_stream<char>&& s) -> future<> {
+            bool first = true;
+            auto res = std::move(_res);
+            co_await s.write("[");
+            co_await do_for_each(res, [&s, &first] (const chunked_stats& v) {
+                return do_for_each(v, [&s, &first] (const task_stats& stats) {
+                    auto fut = first ? make_ready_future<>() : s.write(", ");
+                    first = false;
+                    return fut.then([&s, &stats]() {
+                        tm::task_stats ts;
+                        ts.task_id = stats.task_id;
+                        ts.state = stats.state;
```

Comment:
why do you need to convert it?
Is it a different `task_stats` struct?

Reply:
`tm::task_stats` is a generated struct that has `to_json\(\)` method defined. Though `utils::chunked_vector<T>` requires that "`T` must be nothrow move constructible" which `tm::task_stats` does not satisfy.
Thus I defined `task_stats` struct to use it inside vector and then convert it to `tm::task_stats` to  use `to_json\(\)` method.

Reply:
I see. Let's add a comment about that them, and here, I think that the code can be simplified a bit sing designated struct initializers, like:
```c++
auto ts = tm::task_stats{
    .task_id = stats.task_id;
    .state = stats.state;
};
```